### PR TITLE
genre search functionality

### PR DESCRIPTION
**Pull request by whatdoineed2do**
Add endpoints to:
* list all `genres` known to server
* search by `genres`, returning all albums matching a `genre`

modeled heavily on existing `album` code.

Functionality would be coupled with update to UI that would allow users to queue music by a given genre, such as classical.  Currently, users need to locate and queue individual songs/artists/albums.

~~Missing updates to web ui for which an issue https://github.com/ejurgensen/forked-daapd/issues/558 is logged~~  Web ui (generated/drop-in `htdocs` files) updates on a fork:
~~https://github.com/whatdoineed2do/forked-daapd-web/files/2258629/dist-0cd890b3128c50510c776d455adb0c61a9b3b376.tar.gz~~
~~[dist-search-genre-98ce8de37dfda03a766e5745e5e1219be764545b.tar.gz](https://github.com/ejurgensen/forked-daapd/files/2288253/dist-search-genre-98ce8de37dfda03a766e5745e5e1219be764545b.tar.gz)~~

**EDIT: 2018-08-24**
* branch updated: https://github.com/whatdoineed2do/forked-daapd/tree/search-genre (includes a temporary workaround for https://github.com/ejurgensen/forked-daapd/issues/570)
* webui [dist-search-genre-7d5becba4a8cee675eadbdbadec194dba3386651.tar.gz](https://github.com/ejurgensen/forked-daapd/files/2318320/dist-search-genre-7d5becba4a8cee675eadbdbadec194dba3386651.tar.gz)
* pending to squash commits
* `Q_BROWSE_GENRE`
* adds agreed endpoint, `api/library/genres`
* searches for albums for genres via `api/search`

Please use in conjunction with this build to see intended use cases - web ui code NOT incl in PR)

Upon server functionality is merge, I will submit PR to https://github.com/chme/forked-daapd-web.

comments / feedback?  @chme , @ejurgensen 
Thanks

### CODE REVIEW

**Comment by whatdoineed2do** [comment]
commit https://github.com/ejurgensen/forked-daapd/pull/559/commits/bb26ca39adc8409dbf95b09aa64f33f68681463f is required due to `genres` being outside of strict control of the db (not a generated value).  Therefore the genre name can be anything (ultimately it generates a search/match in the db) so the `json` regex needs to handle.

File: src/httpd_jsonapi.c [hunk 1]
```diff
@@ -2354,6 +2403,146 @@ jsonapi_reply_library_playlist_tracks(struct httpd_request *hreq)
...
+  ret = query_params_limit_set(&query_params, hreq);
+  if (ret < 0)
+    goto error;
+
+  query_params.type = Q_GROUP_ALBUMS;
+  query_params.sort = S_ALBUM;
+  query_params.filter = db_mprintf("(f.genre= '%s')", genre);
+
+  if (media_kind)
+    query_params.filter = db_mprintf("(f.media_kind = %d)", media_kind);
```
Comment (chme) [inline src/httpd_jsonapi.c]: The media_kind filter replaces the filter to fetch a genre (and results in a memory leak). The filters should instead be combined to only fetch the albums of a genre that match the given media_kind.

File: src/httpd_jsonapi.c [hunk 2]
```diff
@@ -2354,6 +2403,146 @@ jsonapi_reply_library_playlist_tracks(struct httpd_request *hreq)
...
+
+  memset(&query_params, 0, sizeof(struct query_params));
+
+  ret = query_params_limit_set(&query_params, hreq);
+  if (ret < 0)
+    goto error;
+
+  query_params.type = Q_GROUP_ALBUMS;
+  query_params.sort = S_ALBUM;
+  query_params.filter = db_mprintf("(f.genre= '%s')", genre);
```
Comment (chme) [inline src/httpd_jsonapi.c]: This needs to be '%q'. Should solve the SQL injection problem with genres that contain '.

**Comment by whatdoineed2do** [comment]
Appreciate the review.

I created a dummy file with the genre `'90s alt` and this works for me locally (https://github.com/ejurgensen/forked-daapd/commit/bb26ca39adc8409dbf95b09aa64f33f68681463f fixed the spaces in names) but we have more of a problem with the genre named `Rock/Pop` since we can't html escape `/`.

Looks like I;ll have to re-factor `api/library/genres/<genre_name>` to `api/library/genres?q="<genre_name"` to handle arbitrary names to solve this.  Try to get this done on the commute to work over the next week if I can't complete it today.

**Comment by ejurgensen** [comment]
I think you should strive to do this with the existing data model and queries. As far I understand what you want to do is return all genres + albums for a particular genre. These are queries that are already supported and implemented for daap (possibly also for rsp), even with optimised indeces, as I recall. So try to use that framework.

There are many ways to group albums and tracks (year, composer, type etc.), and I don't like the idea of having id's for each of these.

**Comment by whatdoineed2do** [comment]
Previous issues fixed, commits squashed.

* no changes to db data model
* endpoints:
`/api/library/genres` - list all known `genres`
`/api/library/genre?name=Rock%2FPop` - search for genre with name to return albums that include requested `genre` (uri encode / escape codes supported, see https://github.com/ejurgensen/forked-daapd/pull/559#issuecomment-412650133)
* webui files updated:
local build support `genre` functionality: https://github.com/ejurgensen/forked-daapd/files/2288253/dist-search-genre-98ce8de37dfda03a766e5745e5e1219be764545b.tar.gz
branch https://github.com/whatdoineed2do/forked-daapd-web/tree/search-genre

Please review, thanks

File: src/logger.c [hunk 3]
```diff
@@ -321,7 +321,7 @@ logger_init(char *file, char *domains, int severity)
       return -1;
     }
 
-  ret = fchown(fileno(logfile), runas_uid, 0);
+  ret = fchown(fileno(logfile), runas_uid, -1);
```
Comment (chme) [inline src/logger.c]: What is the effect of this change? This should at least be a separate commit and better a separate pull request.

**Comment by whatdoineed2do** [comment]
I've been a bit dumb here - I opted for `/api/library/genre?name=...` to overcome the issues with having to escape the genre names having spaces/ampersands/fwd slashes etc.

BUT we can go with `/api/library/genres/<uri encoded/escaped genre name>`

With escaped genre names we can send across `'90s Alternative` / `Rock/Pop' / `Drum & Bass` with no issues so it solves the bug you found last week (my library consists of simple single word genres: Pop, Rock, Classical...)

Does this end point work for you?

On this laptop, using the `api/search?type=album&expression=genre+is+"Alternative` caused the process to lock up.

**Comment by ejurgensen** [comment]
I agree with @chme that it would be good to use the search api for this - i.e. not add a specific endpoint. If it doesn't work then it should be fixed, of course. I also agree with him about the logger.c change. The same goes for admin.html.

@chme is also correct that I think the BROWSE query should be used. Genre is not a group in the data model, and the query with a join on the group table doesn't make sense to me. I suggest first doing the endpoint/UI with existing metadata, and then afterwards expand the BROWSE result if more metadata is required for a better UX.

**Comment by chme** [comment]
@whatdoineed2do In your comment the closing `"` is missing after Alternative. The uri should look like: `api/search?type=album&expression=genre+is+"Alternative"`

I can get forked-daapd to crash, if i am missing the leading `"` (crashes somewhere in the antlr3 code). With a missing trailing `"`, i get a "Bad request" response (correct behavior).

**Comment by ejurgensen** [comment]
/api/library/genres is fine, but as far as I can tell you can use Q_BROWSE_GENRE to query the db, you don't need Q_GROUP_GENRE. You can also later add a /api/library/composers and then use Q_BROWSE_COMPOSERS ... and so on.

I didn't quite understand the two paragraphs after your question, but yes for searching albums belonging to genre it seems good to use the search interface.

**Comment by whatdoineed2do** [comment]
thanks.  Yes, just pushed the most recent fixes to this branch (wasn't 100% ready for review) that now reflects the discussion above:
* `/api/library/genres`
* search is via existing `Q_BROWSE_GROUPS` (not using `GROUPS`)

Need to fix the webui but need to over issues raised in https://github.com/ejurgensen/forked-daapd/issues/570

**Comment by whatdoineed2do** [comment]
I think this is clean for review now - appreciate the patience

* branch updated: https://github.com/whatdoineed2do/forked-daapd/tree/search-genre (includes a temporary workaround for https://github.com/ejurgensen/forked-daapd/issues/570)
* webui [dist-search-genre-7d5becba4a8cee675eadbdbadec194dba3386651.tar.gz](https://github.com/ejurgensen/forked-daapd/files/2318320/dist-search-genre-7d5becba4a8cee675eadbdbadec194dba3386651.tar.gz)
* pending to squash commits
* uses `Q_BROWSE_GENRE` (`db.c` not modified)
* adds agreed endpoint, `api/library/genres`
* searches for albums for genres via `api/search`

File: src/httpd_jsonapi.c [hunk 4]
```diff
@@ -232,6 +232,23 @@ playlist_to_json(struct db_playlist_info *dbpli)
   return item;
 }
 
+static json_object *
+genre_to_json(const char* g_, const struct filecount_info* fci_)
+{
+  json_object *item = NULL;
+  if (g_ == NULL) {
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: Please indent like in the rest of forked-daapd

File: src/httpd_jsonapi.c [hunk 5]
```diff
@@ -232,6 +232,23 @@ playlist_to_json(struct db_playlist_info *dbpli)
   return item;
 }
 
+static json_object *
+genre_to_json(const char* g_, const struct filecount_info* fci_)
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: why the underscores? "genre" and "fci" would be good names

File: src/httpd_jsonapi.c [hunk 6]
```diff
@@ -232,6 +232,23 @@ playlist_to_json(struct db_playlist_info *dbpli)
...
+{
+  json_object *item = NULL;
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: this assignment doesn't seem to do anything

File: src/httpd_jsonapi.c [hunk 7]
```diff
@@ -330,6 +347,7 @@ fetch_artist(const char *artist_id)
   return artist;
 }
 
+
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: ?

File: src/httpd_jsonapi.c [hunk 8]
```diff
@@ -458,6 +476,54 @@ fetch_playlist(const char *playlist_id)
   return playlist;
 }
 
+static int
+fetch_genres(struct query_params *query_params, json_object *items, int *total)
+{
+  json_object *item;
+  int ret = 0;
+  char*  bi;
+  char*  si;
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: why two spaces before bi and si?
naming of bi and si could be better
assignment of ret does nothing

File: src/httpd_jsonapi.c [hunk 9]
```diff
@@ -458,6 +476,54 @@ fetch_playlist(const char *playlist_id)
...
+
+  ret = db_query_start(query_params);
+  if (ret < 0)
+    goto error;
+
+  while (((ret = db_query_fetch_string_sort(query_params, &bi, &si)) == 0) && (bi))
+    {
+      struct filecount_info fci;
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: no mid function declarations in forked-daapd

File: src/httpd_jsonapi.c [hunk 10]
```diff
@@ -2354,6 +2420,73 @@ jsonapi_reply_library_playlist_tracks(struct httpd_request *hreq)
...
+jsonapi_reply_library_genres(struct httpd_request *hreq)
+{
+  time_t db_update;
+  struct query_params query_params;
+  const char *param;
+  enum media_kind media_kind;
+  json_object *reply;
+  json_object *items;
+  int total;
+  int ret = 0;
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: assignment with no effect

**Comment by whatdoineed2do** [comment]
styling fixed.  I still need to squash the commits - we need to agree that the front end functionality is good for you too (a tar file/drop in replacement for the `htdocs` was in the last comment/updated in the OP) in case we need to extend backend further.

**Comment by whatdoineed2do** [comment]
The last bug I mention above is when we use the webui's search page looking for value that doesn't match any genre.  The query from the ui arrives at the server:
`api/search?type=track,artist,album,genre&query=XXX&media_kind=music&limit=3&offset=0`

If XXX exist then this is fine.  However, if XXX is not a known genre then because of uing the `Q_BROWSE_GENRE`, `db_query_start()` drops us into 
```
      default:
        if (qp->type & Q_F_BROWSE)
          query = db_build_query_browse(qp);
```
but this returns NULL from `db_build_query_check() .. db_get_one_int()` because there's no matching rows.

Q - does this mean that the BROWSE searches ALWAYS expects a result?

The web ui's listing of genre uses `api/search?type=albums&expression=genre+is+\"Pop\"` so it's dropping into the `db_build_query_group_albums()`

File: src/httpd_jsonapi.c [hunk 11]
```diff
@@ -234,12 +234,32 @@ playlist_to_json(struct db_playlist_info *dbpli)
   return item;
 }
 
+static json_object *
+genre_to_json(const char* genre, const struct filecount_info* fci)
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: inconsistent asterix placement

File: src/httpd_jsonapi.c [hunk 12]
```diff
@@ -460,6 +480,53 @@ fetch_playlist(const char *playlist_id)
   return playlist;
 }
 
+static int
+fetch_genres(struct query_params *query_params, json_object *items, int *total)
+{
+  json_object *item;
+  int ret;
+  char* genre;
+  char* sort_item;
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: asterix

File: src/httpd_jsonapi.c [hunk 13]
```diff
@@ -460,6 +480,53 @@ fetch_playlist(const char *playlist_id)
...
+
+  while (((ret = db_query_fetch_string_sort(query_params, &genre, &sort_item)) == 0) && (genre))
+    {
+      memset(&fci, 0, sizeof(fci));
+      memset(&qp, 0, sizeof(qp));
+
+      qp.type = Q_COUNT_ITEMS;
+      qp.idx_type = I_NONE;
+      qp.filter = db_mprintf("(f.genre = '%q')", genre);
+      db_query_start(&qp);
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: db_filecount_get starts the query, so you don't need to do it here

File: src/httpd_jsonapi.c [hunk 14]
```diff
@@ -460,6 +480,53 @@ fetch_playlist(const char *playlist_id)
...
+      db_filecount_get(&fci, &qp);
+      free(qp.filter);
+      db_query_end(&qp);
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: also not required

File: src/httpd_jsonapi.c [hunk 15]
```diff
@@ -460,6 +480,53 @@ fetch_playlist(const char *playlist_id)
...
+  if (ret < 0)
+    goto error;
+
+  while (((ret = db_query_fetch_string_sort(query_params, &genre, &sort_item)) == 0) && (genre))
+    {
+      memset(&fci, 0, sizeof(fci));
+      memset(&qp, 0, sizeof(qp));
+
+      qp.type = Q_COUNT_ITEMS;
+      qp.idx_type = I_NONE;
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: is memset of qp and setting consts required inside the loop?

File: src/httpd_jsonapi.c [hunk 16]
```diff
@@ -460,6 +480,53 @@ fetch_playlist(const char *playlist_id)
...
+      qp.filter = db_mprintf("(f.genre = '%q')", genre);
+      db_query_start(&qp);
+      db_filecount_get(&fci, &qp);
```
Comment (ejurgensen) [inline src/httpd_jsonapi.c]: check return value?

**Comment by ejurgensen** [comment]
Thanks for the fixing up, the pr looks fine to me now. As mentioned @chme will also have to check it before merging.

BTW forgot to answer your question: The browse query should work with zero results as well, and if it doesn't it is a bug. I checked the code and it looks ok, db_get_one_int() should get a row because it is a count query (so there should be a row with a "0"). Can I ask you to check again?

**Comment by whatdoineed2do** [comment]
The `json` query for a found/not found genre are identical.  A deeper look we see that the `GROUP BY f.genre(no matching genre found) causes the db handling to be off.

On the sql prompt we can happily run but **note the return on first query** when theres no match, **no rows** are returned.
 ```
sqlite> SELECT COUNT(*) FROM files f WHERE f.disabled = 0 AND (f.genre LIKE '%xxx%') AND f.genre != '' GROUP BY f.genre;
sqlite> SELECT COUNT(*) FROM files f WHERE f.disabled = 0 AND (f.genre LIKE '%Drum%') AND f.genre != '' GROUP BY f.genre;
1
```
without the `GROUP BY` for zero results we get 0 as expected but we need the `GROUP BY` clause.
```
sqlite> SELECT COUNT(*) FROM files f WHERE f.disabled = 0 AND (f.genre LIKE '%xxx%') AND f.genre != '';
0
sqlite> 
```
----
**Proposed Fix**
The fix would be here in `db_get_int_one()`:
```
diff --git a/src/db.c b/src/db.c
index 267d8b73..68e900c8 100644
--- a/src/db.c
+++ b/src/db.c
@@ -1332,12 +1332,18 @@ db_get_one_int(const char *query)
   if (ret != SQLITE_ROW)
     {
       if (ret == SQLITE_DONE)
-       DPRINTF(E_INFO, L_DB, "No matching row found for query: %s\n", query);
+      {
+       DPRINTF(E_INFO, L_DB, "No matching row found for query: %s\n", query);
+        ret = 0;
+      }
       else
+      {
        DPRINTF(E_LOG, L_DB, "Could not step: %s (%s)\n", sqlite3_errmsg(hdl), query);
+        ret = 1;
+      }
 
       sqlite3_finalize(stmt);
-      return -1;
+      return ret;
     }
 
   ret = sqlite3_column_int(stmt, 0);
```
Let me know if this fix is something you are comfortable and I'll  add this to the PR.

File: README_JSON_API.md [hunk 17]
```diff
@@ -1159,6 +1160,151 @@ curl -X GET "http://localhost:3689/api/library/albums/1/tracks"
...
+| items           | array    | Array of [`album`](#album-object) objects |
+| total           | integer  | Total number of albums in the library     |
+| offset          | integer  | Requested offset of the first albums      |
+| limit           | integer  | Requested maximum number of albums        |
+
+
+**Example**
+
+```
+curl -X GET "http://localhost:3689/api/search?type=albums&expression=genre+is+\"Pop\"
```
Comment (ejurgensen) [inline README_JSON_API.md]: The ending quotation mark is missing in these examples

**Comment by ejurgensen** [comment]
In your original post you said the purpose of the pr was:

"Add endpoints to:

    list all genres known to server
    search by genres, returning all albums matching a genre"

But the pr now also extends the search api to return a json genre object, which contains matching genre names plus album, artist and track count. I missed that part in the above conversation, and also when reviewing, I'll admit. I also missed what the purpose of this search is, so I hope you can reiterate - is it to support some sort of future type-ahead search? You wrote above that the web ui only uses type=album, so I am unsure what type=genre is for?

**Comment by whatdoineed2do** [comment]
The work here was to support the following (in order of priority):

* (MUST HAVE) ability for user to see:
.. list of genres for all files on the server and then to play or add the albums directly via genre name (like when you search for album) as returned 
.. to deep dive a return genre listing (from above) to list all albums for a given genre
On the webui (as impl) this functionality is available from `music->tabs` which has been updated to be `queue | artists | albums | genres`
This functionality for use case above is implmented in the backend with `api/library/genres` and `api/search?type=album&expression=genre is xxx`

* (USEFUL) ability to search for genres by name (similar to looking for albums or tracks by name) on the `search` screen and for the matching genres names to be listed; clicking on the genre loads the list of albums for that genre as above in the MUST HAVE.
The webui is implmented to search for a value across album/playlist/artist/genre and this functionality is handled by the code:
`2940  if (strstr(param_type, "genre")) ...` and via the entry `api/search?type=genre...`
This only returns the genre name/total counts for the limited display on the search results screen.
You may wonder why have a search for genre names?  As genre metadata is simply a string, we can have varied and long list of genres and a user may want to search for genres like "Pop" and get a returned set of: "CantoPop", "JPop", "KPop", "MandrinPop", "Pop" on the search screen which they can quickly add to play queue as apposed to scrolling through the entire list of genres (via `music->tabs->genre`)

Let me get some screengrabs when I ahve access to my test system but the tar file updated on the OP (https://github.com/ejurgensen/forked-daapd/files/2318320/dist-search-genre-7d5becba4a8cee675eadbdbadec194dba3386651.tar.gz) can be dropped in place for the existing `htdocs` 
for this tree may explain it easier how these 2 items work

**Comment by ejurgensen** [comment]
I concur with @chme, and also suggest that you move the genre search to another pr. That way we can get the "must have" merged and continue the discussion on the other stuff seperately.

**Comment by whatdoineed2do** [comment]
Updated as per review

I'll send the webui PR shortly

File: src/httpd_jsonapi.c [hunk 18]
```diff
@@ -586,8 +662,10 @@ jsonapi_reply_library(struct httpd_request *hreq)
       DPRINTF(E_LOG, L_WEB, "library: failed to get file count info\n");
     }
 
-  safe_json_add_time_from_string(jreply, "started_at", db_admin_get(DB_ADMIN_START_TIME), true);
-  safe_json_add_time_from_string(jreply, "updated_at", db_admin_get(DB_ADMIN_DB_UPDATE), true);
+  safe_json_add_time_from_string(jreply, "started_at", (s = db_admin_get(DB_ADMIN_START_TIME)), true);
+  free(s);
+  safe_json_add_time_from_string(jreply, "updated_at", (s = db_admin_get(DB_ADMIN_DB_UPDATE)), true);
+  free(s);
```
[the remark this answers is not in the document]
Comment (whatdoineed2do) [reply src/httpd_jsonapi.c]: thanks to everyone's friend, `valgrind`

**Comment by ejurgensen** [comment]
Merged, thanks!
